File: mobile-application/app/src/main/java/com/example/mobile_application/service/ReportsService.java
package com.example.mobile_application.service;

import com.example.mobile_application.dto.StatisticsDTO;
import com.example.mobile_application.dto.UserBasicInfoDTO;

import java.util.List;

import retrofit2.Call;
import retrofit2.http.GET;
import retrofit2.http.Query;

public interface ReportsService {

    @GET("api/reports/statistics")
    Call<StatisticsDTO> getStatistics(
            @Query(value = "userId", encoded = false) Long userId,
            @Query(value = "userType", encoded = false) String userType,
            @Query(value = "filteredUserId", encoded = false) Long filteredUserId,
            @Query(value = "filteredUserType", encoded = false) String filteredUserType,
            @Query(value = "fromDate", encoded = false) String fromDate,
            @Query(value = "toDate", encoded = false) String toDate
    );

    @GET("api/reports/users")
    Call<List<UserBasicInfoDTO>> getAllActiveUsers(
            @Query(value = "excludeUserId", encoded = false) Long excludeUserId
    );

    @GET("api/reports/passengers")
    Call<List<UserBasicInfoDTO>> getAllPassengers();
}

File: client-layer/src/app/services/reports.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../enviroment';

export interface RideDataPoint {
  date: string;
  value: number;
}

export interface Statistics {
  totalRides: number;
  avgRidesPerDay: number;
  ridesData: RideDataPoint[];

  totalKmTraveled: number;
  avgKmPerDay: number;
  kmData: RideDataPoint[];

  totalAmountSpent: number;
  avgAmountPerDay: number;
  amountData: RideDataPoint[];
}

export interface UserBasicInfo {
  id: number;
  email: string;
  firstName: string;
  lastName: string;
  userRole: string; // "PASSENGER" or "DRIVER"
}

@Injectable({
  providedIn: 'root'
})
export class ReportsService {
  private apiUrl = `${environment.apiUrl}/reports`;

  constructor(private http: HttpClient) {}

  getStatistics(userId: number, userType: string, filteredUserId?: number, filteredUserType?: string, fromDate?: string, toDate?: string): Observable<Statistics> {
    let params = new HttpParams()
      .set('userId', userId.toString())
      .set('userType', userType);
    
    if (filteredUserId) {
      params = params.set('filteredUserId', filteredUserId.toString());
    }
    
    if (filteredUserType) {
      params = params.set('filteredUserType', filteredUserType);
    }
    
    if (fromDate) {
      params = params.set('fromDate', fromDate);
    }
    
    if (toDate) {
      params = params.set('toDate', toDate);
    }
    
    return this.http.get<Statistics>(`${this.apiUrl}/statistics`, { params });
  }

  getAllPassengers(): Observable<UserBasicInfo[]> {
    return this.http.get<UserBasicInfo[]>(`${this.apiUrl}/passengers`);
  }

  getAllActiveUsers(excludeUserId: number): Observable<UserBasicInfo[]> {
    const params = new HttpParams().set('excludeUserId', excludeUserId.toString());
    return this.http.get<UserBasicInfo[]>(`${this.apiUrl}/users`, { params });
  }
}
